astropy/units/quantity.py

```diff
@@ -1303,15 +1303,13 @@ def __setitem__(self, i, value):
     # __contains__ is OK
 
     def __bool__(self):
-        """Quantities should always be treated as non-False; there is too much
-        potential for ambiguity otherwise.
+        """This method raises ValueError, since truthiness of quantities is ambiguous,
+        especially for logarithmic units and temperatures. Use explicit comparisons.
         """
-        warnings.warn(
-            "The truth value of a Quantity is ambiguous. "
-            "In the future this will raise a ValueError.",
-            AstropyDeprecationWarning,
+        raise ValueError(
+            "truthiness of quantities is ambiguous, especially for logarithmic units"
```

Comment:
Why make the more general statement that's true for all quantities narrower to make it seem like it only applies to "our" Quantity class?

Reply:
@eerovaher - I'm happy with what @maxnoe has presently. Shall we merge?

Reply:
```suggestion
            f"{type(self).__name__} truthiness is ambiguous, especially for logarithmic units"
```

Reply:
I wasn't referring to potwntial subclasses. I was referring to the general fact in the real world of physival quantities. This is something independent of any implementation.

Reply:
In Python non-empty containers are considered truthy regardless of the truth value of the elements in the container, e.g.
```python
>>> bool([False])
True
```
The fact that this is different for `Quantity` is very much dependent on the fact that `Quantity` is a subclass of `np.ndarray`.

Reply:
I also think that if the error message explicitly included the name of the class then the following could be easier to debug:
```python
>>> from astropy import units as u
>>> as_list = [0 * u.m, 1 * u.m]
>>> as_array = [0, 1] * u.m
>>> as_list == as_array
array([ True,  True])
>>> bool(as_list)
True
>>> bool(as_array)
ValueError
```

Reply:
I'm also totally happy with having the class name in there. If @maxnoe is OK with that too, let's do that and merge.

Reply:
Is this resolve or some kind of follow-up action is needed here?

Reply:
The current warning message explicitly mentions the `Quantity` class. I don't see why it should be changed to avoid mentioning the class entirely, although it would be good if the message would be aware of possible subclasses.